File: SWEA/Unrated/1952. ［모의 SW 역량테스트］ 수영장/［모의 SW 역량테스트］ 수영장.cpp
#include<iostream>
#include<algorithm>
using namespace std;

int T;
int price[4];       // 1일, 1달, 3달, 1년 요금
int plan[13];       // 월별 이용 계획 (1~12월)
int dp[13];

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> T;
    for (int test_case = 1; test_case <= T; test_case++) {

        for (int i = 0; i < 4; i++) cin >> price[i];
        for (int i = 1; i <= 12; i++) cin >> plan[i];

        fill(dp, dp + 13, 1e9);
        dp[0] = 0;

        for (int i = 1; i <= 12; i++) {
            // 1일권으로 i월 처리
            dp[i] = min(dp[i], dp[i - 1] + plan[i] * price[0]);

            // 1달권으로 i월 처리
            dp[i] = min(dp[i], dp[i - 1] + price[1]);

            // 3달권으로 처리 (i >= 3)
            if (i >= 3)
                dp[i] = min(dp[i], dp[i - 3] + price[2]);
            else
                dp[i] = min(dp[i], price[2]); // i < 3일 때 처음부터 3달권 쓸 수 있음
        }

        // 마지막에 1년권과 비교
        int answer = min(dp[12], price[3]);

        cout << "#" << test_case << " " << answer << "\n";
    }
    return 0;
}

File: SWEA/Unrated/1952. ［모의 SW 역량테스트］ 수영장/［모의 SW 역량테스트］ 수영장.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Solution
{
    static int price[];
    static int plan[];
    static int dp[];
	public static void main(String args[]) throws Exception
	{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int T = Integer.parseInt(br.readLine());

		for(int test_case = 1; test_case <= T; test_case++)
		{
            price = new int[4];
            plan = new int[13];
            dp = new int[13];

            StringTokenizer st = new StringTokenizer(br.readLine());
            for(int i = 0; i < 4; i++)  price[i] = Integer.parseInt(st.nextToken());

            st = new StringTokenizer(br.readLine());
            for(int i = 1; i < 13; i++)    plan[i] = Integer.parseInt(st.nextToken());

            for(int i = 0; i < 13; i++) dp[i] = Integer.MAX_VALUE;
            dp[0] = 0;

            for(int i = 1; i < 13; i++) {
                // 일일권으로 i 월 처리
                dp[i] = Math.min(dp[i], dp[i - 1] + plan[i] * price[0]);

                // 한달권으로 i 월 처리
                dp[i] = Math.min(dp[i], dp[i - 1] + price[1]);

                // 세달권으로 처리
                if(i >= 3)  dp[i] = Math.min(dp[i], dp[i - 3] + price[2]);
                else        dp[i] = Math.min(dp[i], price[2]);
            }

            // 마지막으로 1년권과 비교
            int answer = Math.min(dp[12], price[3]);

            System.out.println("#" + test_case + " " + answer);

		}
	}
}